File: examples/csharp/Inventory/src/InventoryLogic.cs
using Angzarr;
using Examples;
using Google.Protobuf;
using Google.Protobuf.WellKnownTypes;
using Serilog;

namespace Angzarr.Examples.Inventory;

public class InventoryLogic : IInventoryLogic
{
    private const int LowStockThreshold = 10;
    private readonly Serilog.ILogger _logger;

    public InventoryLogic(Serilog.ILogger logger)
    {
        _logger = logger.ForContext<InventoryLogic>();
    }

    public InventoryState RebuildState(EventBook? eventBook)
    {
        var state = InventoryState.Empty;

        if (eventBook == null || eventBook.Pages.Count == 0)
            return state;

        foreach (var page in eventBook.Pages)
        {
            if (page.Event == null) continue;
            state = ApplyEvent(state, page.Event);
        }

        return state;
    }

    private InventoryState ApplyEvent(InventoryState state, Any eventAny)
    {
        var typeUrl = eventAny.TypeUrl;

        if (typeUrl.EndsWith("StockInitialized"))
        {
            var evt = eventAny.Unpack<StockInitialized>();
            return state with { ProductId = evt.ProductId, OnHand = evt.Quantity };
        }

        if (typeUrl.EndsWith("StockReceived"))
        {
            var evt = eventAny.Unpack<StockReceived>();
            return state with { OnHand = evt.NewOnHand };
        }

        if (typeUrl.EndsWith("StockReserved"))
        {
            var evt = eventAny.Unpack<StockReserved>();
            var newReservations = new List<Reservation>(state.Reservations) { new(evt.OrderId, evt.Quantity) };
            return state with { Reserved = evt.NewReserved, Reservations = newReservations };
        }

        if (typeUrl.EndsWith("ReservationReleased"))
        {
            var evt = eventAny.Unpack<ReservationReleased>();
            var remainingReservations = state.Reservations.Where(r => r.OrderId != evt.OrderId).ToList();
            return state with { Reserved = evt.NewReserved, Reservations = remainingReservations };
        }

        if (typeUrl.EndsWith("ReservationCommitted"))
        {
            var evt = eventAny.Unpack<ReservationCommitted>();
            var remainingReservations = state.Reservations.Where(r => r.OrderId != evt.OrderId).ToList();
            return state with { OnHand = evt.NewOnHand, Reserved = evt.NewReserved, Reservations = remainingReservations };
        }

        return state;
    }

    public EventBook HandleInitializeStock(InventoryState state, string productId, int quantity)
    {
        if (state.Exists)
            throw CommandValidationException.FailedPrecondition("Inventory already initialized");

        if (string.IsNullOrWhiteSpace(productId))
            throw CommandValidationException.InvalidArgument("Product ID is required");

        if (quantity < 0)
            throw CommandValidationException.InvalidArgument("Quantity cannot be negative");

        _logger.Information("initializing_stock {@Data}", new { product_id = productId, quantity });

        var evt = new StockInitialized
        {
            ProductId = productId,
            Quantity = quantity
        };

        return CreateEventBook(evt);
    }

    public EventBook HandleReceiveStock(InventoryState state, int quantity)
    {
        if (!state.Exists)
            throw CommandValidationException.FailedPrecondition("Inventory not initialized");

        if (quantity <= 0)
            throw CommandValidationException.InvalidArgument("Quantity must be positive");

        var newOnHand = state.OnHand + quantity;

        _logger.Information("receiving_stock {@Data}", new { product_id = state.ProductId, quantity, new_on_hand = newOnHand });

        var evt = new StockReceived
        {
            Quantity = quantity,
            NewOnHand = newOnHand
        };

        return CreateEventBook(evt);
    }

    public EventBook HandleReserveStock(InventoryState state, string orderId, int quantity)
    {
        if (!state.Exists)
            throw CommandValidationException.FailedPrecondition("Inventory not initialized");

        if (string.IsNullOrWhiteSpace(orderId))
            throw CommandValidationException.InvalidArgument("Order ID is required");

        if (quantity <= 0)
            throw CommandValidationException.InvalidArgument("Quantity must be positive");

        if (state.Reservations.Any(r => r.OrderId == orderId))
            throw CommandValidationException.FailedPrecondition("Reservation already exists for order");

        if (quantity > state.Available)
            throw CommandValidationException.FailedPrecondition($"Insufficient stock: available={state.Available}, requested={quantity}");

        var newReserved = state.Reserved + quantity;
        var newAvailable = state.OnHand - newReserved;

        _logger.Information("reserving_stock {@Data}", new { product_id = state.ProductId, order_id = orderId, quantity, new_available = newAvailable });

        var events = new List<IMessage>();

        events.Add(new StockReserved
        {
            OrderId = orderId,
            Quantity = quantity,
            NewReserved = newReserved
        });

        if (newAvailable < LowStockThreshold && state.Available >= LowStockThreshold)
        {
            events.Add(new LowStockAlert
            {
                ProductId = state.ProductId,
                AvailableQuantity = newAvailable,
                Threshold = LowStockThreshold
            });
        }

        return CreateEventBook(events);
    }

    public EventBook HandleReleaseReservation(InventoryState state, string orderId)
    {
        if (!state.Exists)
            throw CommandValidationException.FailedPrecondition("Inventory not initialized");

        var reservation = state.Reservations.FirstOrDefault(r => r.OrderId == orderId)
            ?? throw CommandValidationException.FailedPrecondition("Reservation not found");

        var newReserved = state.Reserved - reservation.Quantity;

        _logger.Information("releasing_reservation {@Data}", new { product_id = state.ProductId, order_id = orderId, quantity = reservation.Quantity, new_reserved = newReserved });

        var evt = new ReservationReleased
        {
            OrderId = orderId,
            Quantity = reservation.Quantity,
            NewReserved = newReserved
        };

        return CreateEventBook(evt);
    }

    public EventBook HandleCommitReservation(InventoryState state, string orderId)
    {
        if (!state.Exists)
            throw CommandValidationException.FailedPrecondition("Inventory not initialized");

        var reservation = state.Reservations.FirstOrDefault(r => r.OrderId == orderId)
            ?? throw CommandValidationException.FailedPrecondition("Reservation not found");

        var newOnHand = state.OnHand - reservation.Quantity;
        var newReserved = state.Reserved - reservation.Quantity;

        _logger.Information("committing_reservation {@Data}", new { product_id = state.ProductId, order_id = orderId, quantity = reservation.Quantity, new_on_hand = newOnHand });

        var evt = new ReservationCommitted
        {
            OrderId = orderId,
            Quantity = reservation.Quantity,
            NewOnHand = newOnHand,
            NewReserved = newReserved
        };

        return CreateEventBook(evt);
    }

    private static EventBook CreateEventBook(IMessage evt)
    {
        return CreateEventBook(new List<IMessage> { evt });
    }

    private static EventBook CreateEventBook(List<IMessage> events)
    {
        var book = new EventBook();

        for (var i = 0; i < events.Count; i++)
        {
            var page = new EventPage
            {
                Num = i,
                Event = Any.Pack(events[i]),
                CreatedAt = Timestamp.FromDateTime(DateTime.UtcNow)
            };
            book.Pages.Add(page);
        }

        return book;
    }
}

File: examples/cpp/inventory/src/inventory_logic.cc
#include "inventory_logic.hpp"
#include "validation_error.hpp"
#include <chrono>

namespace inventory {

using namespace angzarr;

InventoryState InventoryLogic::rebuild_state(const angzarr::EventBook* event_book) {
    InventoryState state;
    if (!event_book || event_book->pages().empty()) return state;

    for (const auto& page : event_book->pages()) {
        if (page.has_event()) {
            state = apply_event(std::move(state), page.event());
        }
    }
    return state;
}

examples::StockInitialized InventoryLogic::handle_initialize_stock(
    const InventoryState& state, const std::string& product_id, int32_t quantity) {
    if (state.exists()) throw ValidationError::failed_precondition("Inventory already exists");
    if (product_id.empty()) throw ValidationError::invalid_argument("Product ID is required");
    if (quantity < 0) throw ValidationError::invalid_argument("Quantity cannot be negative");

    examples::StockInitialized event;
    event.set_product_id(product_id);
    event.set_initial_quantity(quantity);
    return event;
}

examples::StockReceived InventoryLogic::handle_receive_stock(
    const InventoryState& state, int32_t quantity, const std::string& reference) {
    if (!state.exists()) throw ValidationError::failed_precondition("Inventory does not exist");
    if (quantity <= 0) throw ValidationError::invalid_argument("Quantity must be positive");

    examples::StockReceived event;
    event.set_quantity(quantity);
    event.set_reference(reference);
    event.mutable_received_at()->set_seconds(
        std::chrono::system_clock::to_time_t(std::chrono::system_clock::now()));
    return event;
}

examples::StockReserved InventoryLogic::handle_reserve_stock(
    const InventoryState& state, const std::string& order_id, int32_t quantity) {
    if (!state.exists()) throw ValidationError::failed_precondition("Inventory does not exist");
    if (order_id.empty()) throw ValidationError::invalid_argument("Order ID is required");
    if (quantity <= 0) throw ValidationError::invalid_argument("Quantity must be positive");
    if (state.reservations.find(order_id) != state.reservations.end()) {
        throw ValidationError::failed_precondition("Reservation already exists for this order");
    }
    if (quantity > state.available()) throw ValidationError::failed_precondition("Insufficient stock");

    examples::StockReserved event;
    event.set_order_id(order_id);
    event.set_quantity(quantity);
    event.set_available_after(state.available() - quantity);
    return event;
}

examples::ReservationReleased InventoryLogic::handle_release_reservation(
    const InventoryState& state, const std::string& order_id) {
    if (!state.exists()) throw ValidationError::failed_precondition("Inventory does not exist");
    auto it = state.reservations.find(order_id);
    if (it == state.reservations.end()) {
        throw ValidationError::failed_precondition("No reservation found for this order");
    }

    examples::ReservationReleased event;
    event.set_order_id(order_id);
    event.set_quantity_released(it->second.quantity);
    return event;
}

examples::ReservationCommitted InventoryLogic::handle_commit_reservation(
    const InventoryState& state, const std::string& order_id) {
    if (!state.exists()) throw ValidationError::failed_precondition("Inventory does not exist");
    auto it = state.reservations.find(order_id);
    if (it == state.reservations.end()) {
        throw ValidationError::failed_precondition("No reservation found for this order");
    }

    examples::ReservationCommitted event;
    event.set_order_id(order_id);
    event.set_quantity_committed(it->second.quantity);
    return event;
}

InventoryState InventoryLogic::apply_event(InventoryState state, const google::protobuf::Any& event) {
    const auto& type_url = event.type_url();

    if (type_url.find("StockInitialized") != std::string::npos) {
        examples::StockInitialized e;
        event.UnpackTo(&e);
        state.product_id = e.product_id();
        state.on_hand = e.initial_quantity();
    } else if (type_url.find("StockReceived") != std::string::npos) {
        examples::StockReceived e;
        event.UnpackTo(&e);
        state.on_hand += e.quantity();
    } else if (type_url.find("StockReserved") != std::string::npos) {
        examples::StockReserved e;
        event.UnpackTo(&e);
        state.reserved += e.quantity();
        state.reservations[e.order_id()] = {e.order_id(), e.quantity()};
    } else if (type_url.find("ReservationReleased") != std::string::npos) {
        examples::ReservationReleased e;
        event.UnpackTo(&e);
        auto it = state.reservations.find(e.order_id());
        if (it != state.reservations.end()) {
            state.reserved -= it->second.quantity;
            state.reservations.erase(it);
        }
    } else if (type_url.find("ReservationCommitted") != std::string::npos) {
        examples::ReservationCommitted e;
        event.UnpackTo(&e);
        auto it = state.reservations.find(e.order_id());
        if (it != state.reservations.end()) {
            state.on_hand -= it->second.quantity;
            state.reserved -= it->second.quantity;
            state.reservations.erase(it);
        }
    }

    return state;
}

}  // namespace inventory
